src/formats/CIF.cpp

```diff
@@ -0,0 +1,364 @@
+
+        if (group_pdb != atom_site_map_.end()) {
+            atom.set("is_hetatm", line_split[group_pdb->second] != "ATOM");
+        }
+
+        if (label_alt_id != atom_site_map_.end() &&
+            line_split[label_alt_id->second] != ".") {
+            atom.set("altloc", line_split[label_alt_id->second]);
+        }
+        
+        if (formal_charge != atom_site_map_.end()) {
+            atom.set_charge(cif_to_double(line_split[formal_charge->second]));
+        }
+
+        if (!uses_fract_) {
+            auto x = cif_to_double(line_split[cartn_x->second]);
+            auto y = cif_to_double(line_split[cartn_y->second]);
+            auto z = cif_to_double(line_split[cartn_z->second]);
+
+            frame.add_atom(std::move(atom), Vector3D(x, y, z));
+        } else {
+            auto u = cif_to_double(line_split[fract_x->second]);
+            auto v = cif_to_double(line_split[fract_y->second]);
+            auto w = cif_to_double(line_split[fract_z->second]);
+
+            auto vector = cell_.matrix() * Vector3D(u, v, w);
+            frame.add_atom(std::move(atom), vector);
+        }
+
+        position = file_->tellg();
+        line = file_->readline();
+
+        if (label_comp_id == atom_site_map_.end() ||
+            label_asym_id == atom_site_map_.end()) {
+            continue;
+        }
+
+        auto atom_id = frame.size() - 1;
+        size_t resid;
+        auto resid_text = line_split[label_seq_id->second];
+
+        try {
+            if (resid_text == ".") { // In this case, we need to use the entity id
+                resid = std::stoul(line_split[label_entity_id->second]);
+            } else {
+                resid = std::stoul(line_split[label_seq_id->second]);
+            }
+        } catch (std::invalid_argument& e) {
+            throw format_error("Invalid CIF residue or entity numeric: {}", e.what());
+        }
+
+        auto chainid = line_split[label_asym_id->second];
+
+        if (residues_.find({chainid, resid}) == residues_.end()) {
+
+            auto name = line_split[label_comp_id->second];
+            Residue residue(std::move(name), resid);
+            residue.add_atom(atom_id);
+
+            // This will be save as a string... on purpose to match MMTF
+            if (label_asym_id != atom_site_map_.end()) {
+                residue.set("chainid", chainid);
+            }
+
+            if (auth_asym_id != atom_site_map_.end()) {
+                residue.set("chainname", line_split[auth_asym_id->second]);
+            }
+
+            residues_.insert({{chainid, resid}, residue});
+        } else {
+            // Just add this atom to the residue
+            residues_.at({chainid, resid}).add_atom(atom_id);
+        }
+    }
+
+    file_->seekg(position);
+
+    for (const auto& residue: residues_) {
+        frame.add_residue(residue.second);
+    }
+
+    // Cross format talk! Forgive me!
+    PDBFormat::link_standard_residue_bonds(frame);
```

Comment:
I like the comment ^-^

On a side note, this is not valid for usual CIF files, only biomolecules ones from the PDB archive and consorts. I though there where using the .mmcif extension for there own extensions to the format

Reply:
They use .cif, quite confusing.

Reply:
Is there some kind of RCSB specific key we could check for before applying this ?

Reply:
I already use the `_atom_site.pdbx_pdb_model_num` key to identify a CIF from the PDB. This key is present in 94% of all CIF files from the PDB, so I could use this again.

On another note, I am wondering if this should be relabeled as mmCIF support.  The CIF write code I am working writes cartesian coordinates (I mirror PyMOL and RCSB here).

Reply:
Yeah, both sounds good !

Reply:
I checked OpenBabel as well and it is done in a similar manner. They implement mmCIF and CIF as different formats (the read code is similar, but registered differently). The write code is different for the two formats, one writes cartesian and the other fractional.  This is probably the best way to proceed.  It does not matter to me as the 'mmCIF' parser here works for the 'CIF' files I am working with anyway.